pkg/dns/ibm/dns_test.go

```diff
@@ -0,0 +1,273 @@
+			}
+
+		})
+	}
+}
+
+func TestCreateOrUpdate(t *testing.T) {
+
+	provider := &Provider{}
+
+	dnsService, err := dnsclient.NewFake()
+
+	provider.dnsServices = make(map[string]dnsclient.DnsClient)
+
+	zone := configv1.DNSZone{
+		ID: "zoneID",
+	}
+
+	provider.dnsServices[zone.ID] = dnsService
+
+	if err != nil {
+		t.Error("failed to create fakeClient")
+		return
+	}
+
+	testCases := []struct {
+		desc                         string
+		DNSName                      string
+		recordedCall                 string
+		listAllDnsRecordsInputOutput dnsclient.ListAllDnsRecordsInputOutput
+		updateDnsRecordInputOutput   dnsclient.UpdateDnsRecordInputOutput
+		expectedErr                  bool
+	}{
+		{
+			desc:         "happy path",
+			DNSName:      "testUpdate",
+			recordedCall: "PUT",
+			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
+				OutputError:      nil,
+				OutputStatusCode: http.StatusOK,
+			},
+			updateDnsRecordInputOutput: dnsclient.UpdateDnsRecordInputOutput{
+				InputId:          "testUpdate",
+				OutputError:      nil,
+				OutputStatusCode: http.StatusOK,
+			},
+			expectedErr: false,
+		},
+		{
+			desc:         "listFail",
+			DNSName:      "testUpdate",
+			recordedCall: "PUT",
+			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
+				OutputError:      nil,
+				OutputStatusCode: http.StatusNotFound,
+			},
+			expectedErr: false,
+		},
+		{
+			desc:         "listFailError",
+			DNSName:      "testUpdate",
+			recordedCall: "PUT",
+			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
+				OutputError:      errors.New("Error in ListAllDnsRecords"),
+				OutputStatusCode: http.StatusRequestTimeout,
+			},
+			expectedErr: true,
+		},
+		{
+			desc:         "updateError",
+			DNSName:      "testUpdate",
+			recordedCall: "PUT",
+			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
+				OutputError:      nil,
+				OutputStatusCode: http.StatusOK,
+			},
+			updateDnsRecordInputOutput: dnsclient.UpdateDnsRecordInputOutput{
+				InputId:          "testUpdate",
+				OutputError:      errors.New("Error in UpdateDnsRecord"),
+				OutputStatusCode: http.StatusRequestTimeout,
+			},
+			expectedErr: true,
+		},
+		{
+			desc:         "empty DNSName",
+			DNSName:      "",
+			recordedCall: "",
+			expectedErr:  true,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+
+			record := iov1.DNSRecord{
+				Spec: iov1.DNSRecordSpec{
+					DNSName:    tc.DNSName,
+					RecordType: iov1.ARecordType,
+					Targets:    []string{"11.22.33.44"},
+					RecordTTL:  120,
+				},
+			}
+
+			dnsService.ListAllDnsRecordsInputOutput = tc.listAllDnsRecordsInputOutput
+
+			dnsService.UpdateDnsRecordInputOutput = tc.updateDnsRecordInputOutput
+
+			err = provider.createOrUpdateDNSRecord(&record, zone)
+
+			if tc.expectedErr {
+				if err == nil {
+					t.Error("Expected error, but err is nil")
+				}
+			}
```

Comment:
Check `!tc.expectedErr && err != nil`: 
```go
			if !tc.expectedErr && err != nil {
				t.Errorf("Expected nil err, got %v", err)
			}
```
If you do that, I think you'll find that my other comment about checking for "not found" errors in the create/update logic has merit.  * grin *.

Reply:
Hm, on second look, maybe the "not found" errors aren't relevant.  This check causes the following failure:
```
--- FAIL: TestCreateOrUpdate (0.00s)
    --- FAIL: TestCreateOrUpdate/listFail (0.00s)
        dns_test.go:261: Expected nil err, got createOrUpdateDNSRecord: failed to update the dns record: updateDnsRecord: inputs don't match
```
That can be fixed by adding `updateDnsRecordInputOutput` to the test case:
```go
		{
			desc:         "listFail",
			DNSName:      "testUpdate",
			recordedCall: "PUT",
			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
				OutputError:      nil,
				OutputStatusCode: http.StatusNotFound,
			},
			updateDnsRecordInputOutput: dnsclient.UpdateDnsRecordInputOutput{
				InputId:          "testUpdate",
				OutputError:      nil,
				OutputStatusCode: http.StatusOK,
			},
			expectedErr: false,
		},
```

Reply:
Ahh, I think the test case should actually look like this:
```go
		{
			desc:         "listFail",
			DNSName:      "testUpdate",
			recordedCall: "PUT",
			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
				OutputError:      errors.New("Error in ListAllDnsRecords"),
				OutputStatusCode: http.StatusNotFound,
			},
			updateDnsRecordInputOutput: dnsclient.UpdateDnsRecordInputOutput{
				InputId:          "testUpdate",
				OutputError:      nil,
				OutputStatusCode: http.StatusOK,
			},
			expectedErr: false,
		},
```
Then I get the following failure:
```
--- FAIL: TestCreateOrUpdate (0.00s)
    --- FAIL: TestCreateOrUpdate/listFail (0.00s)
        dns_test.go:266: Expected nil err, got createOrUpdateDNSRecord: failed to list the dns record: Error in ListAllDnsRecords
```
But if I fix `createOrUpdateDNSRecord` as follows:
```go
		result, response, err := dnsService.ListAllDnsRecords(listOpt)
		if err != nil {
			if response.StatusCode != http.StatusNotFound {
				return fmt.Errorf("createOrUpdateDNSRecord: failed to list the dns record: %w", err)
			}
			continue
		}
```
Then the test passes again.